src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -284,18 +284,20 @@ public bool Equals(JsonSerializerOptions? left, JsonSerializerOptions? right)
                     left._typeInfoResolver == right._typeInfoResolver &&
                     CompareLists(left._converters, right._converters);
 
-                static bool CompareLists<TValue>(ConfigurationList<TValue> left, ConfigurationList<TValue> right)
+                static bool CompareLists<TValue>(ConfigurationList<TValue>? left, ConfigurationList<TValue>? right)
                     where TValue : class?
                 {
-                    int n;
-                    if ((n = left.Count) != right.Count)
+                    int leftCount = left is null ? 0 : left.Count;
```

Comment:
Possibly we could add equals and hash code to our implementations of converters and make them return trues when types match? That would make it relatively more likely to be true

Reply:
> Won't this converters check realistically always be false?

It would almost always be false, I wouldn't expect users to cache converters if they're not caching options.

> Possibly we could add equals and hash code to our implementations of converters and make them return trues when types match? That would make it relatively more likely to be true

It might also be false because some converters accept parameters. We want to avoid false positives under all circumstances since it would lead to observable bugs -- on the other hand we tolerate false negatives since they can only result in reduced performance.